### ISSUE TITLE
main thread panic in Result::unwrap (mainnet)

### ISSUE THREAD

**Issue by damian0815** [tzConnectBerlin/que-pasa#42]
I'm getting the following error running quepasa on mainnet (hangzhounet seems to be fine):

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed to get level_json for level=2062509
```

Full output:

```
damian@d-mba2 que-pasa % RUST_BACKTRACE=1 NODE_URL=https://mainnet.api.tez.ie DATABASE_NAME=quepasa-mainnet DATABASE_URL=[[omitted]] BCD_URL=https://api.better-call.dev/v1 BCD_NETWORK=mainnet NETWORK=mainnet cargo run -- --bcd-enable --contracts [[omitted]]

    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/que-pasa --bcd-enable --contracts [[omitted]]
[2022-01-26T12:06:37Z INFO  que_pasa] running for contracts: [
        ContractID {
...
        },
    ]
[2022-01-26T12:06:39Z INFO  que_pasa::sql::db] querying db to check for any missing levels of ewy_mainnet..
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed to get level_json for level=2062509

Caused by:
    Stream error in the HTTP/2 framing layer (verbose: HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)) (curl status code: 92)', src/main.rs:155:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1298:23
   4: que_pasa::main
             at ./src/main.rs:148:5
   5: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by RKlompUU** [comment]
Hi Damian, does it happen consistently with this block specifically? If no, does it happen from time to time?

On first impression this seems like a Tezos node that is not behaving correctly (the communication stream not being closed correctly hints at it, which presumably is happening because something else goes wrong - inside the node - during the RPC call to the node).

Personally tried to index this block with a public Tezos node (https://mainnet-tezos.giganode.io) and it went fine.

Perhaps there are ways that Que Pasa could deal with this scenario better, but I think I need some more information from you in order to go forward.

**Comment by damian0815** [comment]
ok, so indexing was able to resume with https://mainnet-tezos.giganode.io - i let it run from 2062509 to 2062535, then switched back to https://mainnet.api.tez.ie and now it's working fine. so it does seem like it was just the combination of block 2062509 / https://mainnet.api.tez.ie . i don't know enough about postgresql to be able to roll back the db to try that block again though :/

**Comment by damian0815** [comment]
fwiw i'm able to reproduce the curl error on the command line with `curl -vv https://mainnet.api.tez.ie/chains/main/blocks/2062509 > /tmp/2062509.block`:
```
{ [871 bytes data]
 17 7735k   17 1330k    0     0  78413      0  0:01:41  0:00:17  0:01:24 90884* HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
* stopped the pause stream!
 18 7735k   18 1423k    0     0  80043      0  0:01:38  0:00:18  0:01:20   97k
* Connection #0 to host mainnet.api.tez.ie left intact
curl: (92) HTTP/2 stream 0 was not closed cleanly: INTERNAL_ERROR (err 2)
```

**Comment by RKlompUU** [comment]
Interesting. So it is this mainnet.api.tez node that is causing trouble somehow.

It looks like Que Pasa cannot do much better here. This is not a transient error, if it happens it will happen again when attempting to contact the same node for the same block. So, adding it to that class of errors and retrying the RPC call will not help at all here.

The only way that Que Pasa could maybe do better is have a fallback node for such kind of errors; then it can retry the RPC call on this fallback node.

Will leave this issue open so that discussion can continue if there's more to be said. And planning to modify the node url parameter to optionally be an array of node urls, the first being the primary and the rest being fallback nodes for any non-transient errors.

**Comment by damian0815** [comment]
ping @RKlompUU - I'm seeing error messages like
`Failure when receiving data from the peer (verbose: OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104) (curl status code: 56)', src/main.rs:159:32` and `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed to deserialize block json`. This is using the giganode.io RPC nodes.

These errors tend to disappear on restart. Ideally, quepasa should always back off and retry error'd calls like this.

**Comment by damian0815** [comment]
also DB connection issues could be treated as transient - at the moment when my frontend gets stressed, the backend can run out of connections to the postgresql instance, and as soon as this happens que-pasa simply terminates:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed to connect to the db "
```

**Comment by RKlompUU** [comment]
@damian0815 

here's the PR: https://github.com/tzConnectBerlin/que-pasa/pull/45

The only thing I haven't been able to address is the "thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: failed to deserialize block json". Unfortunately I think I'd need a bit more context in order to look into this. I have to know where exactly this `unwrap()` is called, and most likely a bit more context besides that as well. If it is not a problem for you, could you run with rust nightly mode and with `RUST_BACKTRACE` environment variable defined to `1`? Alike this: `cargo +nightly run -- ...` (substituting `...` with your own que pasa arguments).

I'll leave the PR open for a bit, until tomorrow morning or so. But probably we wont be able to address this json deserialization issue by then. I think that issue will wait for a later PR, so we can get what's been done already into a new version directly.

**Comment by RKlompUU** [comment]
Solved _I think_ all of your remarks. Was merged into main with PR #45, release version v1.1.1.

Closing this issue, should I have missed anything then I think it's best to open a new Issue rather than continuing here.

**Comment by damian0815** [comment]
thanks @RKlompUU - i've scheduled an issue to try it out on the Endless Ways staging indexer, will let you know how it works out when i've done that.
